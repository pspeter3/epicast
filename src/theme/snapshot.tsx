import * as React from "react";
import TestRenderer from "react-test-renderer";

export const snapshot = (name: string, SFC: React.SFC<{}>): void => {
    describe(name, () => {
        it(`should match snapshot`, () => {
            const component = TestRenderer.create(<SFC>Snapshot</SFC>);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
};
