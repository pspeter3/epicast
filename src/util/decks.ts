import { Deck } from "../core/types";
import { size } from "./stacks";

export const unique = (deck: Deck): string[] =>
    Object.keys(
        deck.reduce(
            (set, stack) => {
                return Object.keys(stack).reduce((collection, key) => {
                    collection[key] = null;
                    return collection;
                }, set);
            },
            {} as Record<string, null>,
        ),
    );

export const total = (deck: Deck): number => deck.reduce((sum, stack) => sum + size(stack), 0);
