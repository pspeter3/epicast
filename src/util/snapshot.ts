import { shallow } from "enzyme";
import * as React from "react";

export type Factory = () => React.ReactElement<any>;

export const snapshot = (factory: Factory): (() => void) => {
    return () => {
        expect(shallow(factory())).toMatchSnapshot();
    };
};

export const snapshotSuite = (name: string, factory: Factory): void => {
    describe(name, () => {
        it("should match snapshot", snapshot(factory));
    });
};
