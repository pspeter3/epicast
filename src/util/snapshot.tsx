import * as React from "react";
import TestRenderer from "react-test-renderer";

export type Factory = () => React.ReactElement<any>;

export const snapshot = (factory: Factory): (() => void) => {
    return () => {
        const component = TestRenderer.create(factory());
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    };
};

export const snapshotSuite = (name: string, factory: Factory): void => {
    describe(name, () => {
        it("should match snapshot", snapshot(factory));
    });
};
