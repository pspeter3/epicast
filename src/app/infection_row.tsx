import * as React from "react";
import { classNames } from "../theme/css";
import { NumericInput } from "../theme/numeric_input";
import { Display, JustifyContent, Padding } from "../theme/tailwind";

export interface Props {
    city: string;
    value: number;
    onChange: (city: string, value: number) => void;
}

export class InfectionRow extends React.PureComponent<Props, {}> {
    public static displayName = "InfectionRow";

    public render() {
        const { city, value } = this.props;
        return (
            <div className={classNames(Display.Flex, JustifyContent.Between)}>
                <label className={classNames(Padding.A3)}>{city}</label>
                <NumericInput value={value} onChange={this._onChange} />
            </div>
        );
    }

    private _onChange = (value: number): void => {
        this.props.onChange(this.props.city, value);
    };
}
