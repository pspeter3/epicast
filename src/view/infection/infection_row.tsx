import * as React from "react";
import { IconButton } from "../../theme/buttons";
import { TargetIcon } from "../../theme/icons";
import { Checkbox } from "../../theme/inputs";
import { Row } from "../../theme/layout";
import { Text } from "../../theme/typography";

export interface Props {
    city: string;
    checked?: boolean;
    onToggle: (city: string) => void;
    onEpidemic?: (city: string) => void;
}

export class InfectionRow extends React.PureComponent<Props, {}> {
    public render() {
        const { city, checked, onEpidemic } = this.props;
        return (
            <Row>
                <Checkbox aria-checked={checked} onClick={this._onToggle} />
                <Text>{city}</Text>
                {onEpidemic && (
                    <IconButton onClick={this._onEpidemic}>
                        <TargetIcon />
                    </IconButton>
                )}
            </Row>
        );
    }

    private _onToggle = () => {
        this.props.onToggle(this.props.city);
    };

    private _onEpidemic = () => {
        this.props.onEpidemic!(this.props.city);
    };
}
