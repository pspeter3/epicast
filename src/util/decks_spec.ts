import { total, unique } from "./decks";

const PRIMARY = "San Francisco";
const SECONDARY = "Los Angeles";

describe("decks", () => {
    describe("unique", () => {
        it("should return the unique cards", () => {
            expect(
                unique([
                    { [PRIMARY]: 1 },
                    {
                        [PRIMARY]: 2,
                        [SECONDARY]: 1,
                    },
                ]),
            ).toEqual([PRIMARY, SECONDARY]);
        });
    });

    describe("total", () => {
        it("should return the exact count", () => {
            expect(
                total([
                    { [PRIMARY]: 1 },
                    {
                        [PRIMARY]: 2,
                        [SECONDARY]: 1,
                    },
                ]),
            ).toBe(4);
        });
    });
});
