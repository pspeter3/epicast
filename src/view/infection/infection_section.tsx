import * as React from "react";
import { Stack } from "../../core/types";
import { Section } from "../../theme/layout";
import { value } from "../../util/stacks";
import { InfectionRow } from "./infection_row";

export interface Props {
    index: number;
    stack: Stack;
    checked: Stack;
    onToggle: (index: number, city: string) => void;
    onEpidemic?: (city: string) => void;
}

export class InfectionSection extends React.PureComponent<Props, {}> {
    public render() {
        const { stack, checked, onEpidemic } = this.props;
        return (
            <Section>
                {Object.keys(stack)
                    .sort()
                    .reduce(
                        (children, city) => {
                            const count = stack[city];
                            const numChecked = value(checked, city);
                            for (let i = 0; i < count; i++) {
                                children.push(
                                    <InfectionRow
                                        key={`${city}${i}`}
                                        city={city}
                                        checked={i < numChecked}
                                        onToggle={this._onToggle}
                                        onEpidemic={onEpidemic}
                                    />,
                                );
                            }
                            return children;
                        },
                        [] as Array<React.ReactElement<{}>>,
                    )}
            </Section>
        );
    }

    private _onToggle = (city: string) => {
        this.props.onToggle(this.props.index, city);
    };
}
