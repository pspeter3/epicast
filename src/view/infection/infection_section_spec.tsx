import { shallow } from "enzyme";
import * as React from "react";
import { snapshot } from "../../util/snapshot";
import { InfectionRow } from "./infection_row";
import { InfectionSection } from "./infection_section";

const PRIMARY = "San Francisco";
const SECONDARY = "Los Angeles";

describe("InfectionSection", () => {
    describe("snapshots", () => {
        it(
            "should match snapshot with checked cities",
            snapshot(() => (
                <InfectionSection
                    index={0}
                    stack={{ [PRIMARY]: 2, [SECONDARY]: 1 }}
                    checked={{ [PRIMARY]: 1 }}
                    onToggle={jest.fn()}
                />
            )),
        );
        it(
            "should match snapshot with epidemic",
            snapshot(() => (
                <InfectionSection
                    index={0}
                    stack={{ [PRIMARY]: 2, [SECONDARY]: 1 }}
                    checked={{}}
                    onToggle={jest.fn()}
                    onEpidemic={jest.fn()}
                />
            )),
        );
    });

    describe("handlers", () => {
        it("should handle toggle", () => {
            const onToggle = jest.fn();
            const tree = shallow(
                <InfectionSection
                    index={0}
                    stack={{ [PRIMARY]: 2, [SECONDARY]: 1 }}
                    checked={{}}
                    onToggle={onToggle}
                />,
            );
            tree
                .find(InfectionRow)
                .first()
                .props().onToggle!(PRIMARY);
            expect(onToggle).toHaveBeenCalledWith(0, PRIMARY);
        });
    });
});
