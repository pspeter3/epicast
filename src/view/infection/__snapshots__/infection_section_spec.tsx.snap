// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfectionSection snapshots should match snapshot with checked cities 1`] = `
<Component>
  <InfectionRow
    checked={false}
    city="Los Angeles"
    key="Los Angeles0"
    onToggle={[Function]}
  />
  <InfectionRow
    checked={true}
    city="San Francisco"
    key="San Francisco0"
    onToggle={[Function]}
  />
  <InfectionRow
    checked={false}
    city="San Francisco"
    key="San Francisco1"
    onToggle={[Function]}
  />
</Component>
`;

exports[`InfectionSection snapshots should match snapshot with epidemic 1`] = `
<Component>
  <InfectionRow
    checked={false}
    city="Los Angeles"
    key="Los Angeles0"
    onEpidemic={[MockFunction]}
    onToggle={[Function]}
  />
  <InfectionRow
    checked={false}
    city="San Francisco"
    key="San Francisco0"
    onEpidemic={[MockFunction]}
    onToggle={[Function]}
  />
  <InfectionRow
    checked={false}
    city="San Francisco"
    key="San Francisco1"
    onEpidemic={[MockFunction]}
    onToggle={[Function]}
  />
</Component>
`;
